### PR TITLE
fix: correctly parse browser as config as string

### PR DESCRIPTION
**Pull request by lukekarrys**
### <samp>🤖 Generated by Copilot at 5ef6c04</samp>

This pull request introduces a new `mock-globals` package to the npm cli project, which provides a way to mock the global objects and variables used by the npm cli code. It also updates the code, tests, documentation, and configuration files of the npm cli project to use the `mock-globals` package and to improve the quality and compatibility of the code. Additionally, it adds a new GitHub Actions workflow to run linting and testing for the `mock-globals` package, and a new `.eslintrc.local.js` file to enforce es6 syntax and restrict require statements for certain files.
​

### <samp>🤖 Generated by Copilot at 5ef6c04</samp>

> _We're the crew of the npm cli_
> _We work on code and docs with glee_
> _We mock the globals and lint the files_
> _And test them all with `tap` and `tmock`_

### <samp>🤖 Generated by Copilot at 5ef6c04</samp>

*  Move engine validation logic to a separate file and require it from `index.js` ([link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-e727e4bdf3657fd1d798edcd6b099d6e092f8573cba266154583a746bba0f346L2-R2), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-e2954b558f2aa82baff0e30964490d12942e0e251c1aa56c3294de6ec67b7cf5R18))
*  Add `mock-globals` package to provide a mock implementation of global variables for testing ([link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-62bf6befbad2ef212a49a0096fb2a0de97f63bb7b26fcc3ff10772c5fc7f68f8R1-R17), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-6df241d8da91be5a875147caddcc20e1cdcf81fbff333b9c7ac03114321c60a0R1-R21), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-f24f5e8b4df8e3de3f4f8639183476215831d61f8eab8fbfdcefb5a14e51c96fR1-R50), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-978ca80f731e245f34e8cbedb33d9c79cbe17774adfd870e78ad3073133a791aL2-R2), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eR123), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eR217-R218), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eR541), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eR687-R689), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L9-R10), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R198), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L247-R253), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-728d7692795781aa584471a5f5b12b11c03690c9d7189c43d4475ee32ea3638bR1-R94), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR13), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-b803fcb7f17ed9235f1e5cb1fcd2f5d3b2838429d4368ae4c57ce4436577f03fR23))
*  Use `npmcli-config` package to handle npm config options and constants ([link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-2e7bd92a74f1fb3fe469008027f9ceb4c81082cb8b234ebd3c1b33c20b1af892L5-R14), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-2e7bd92a74f1fb3fe469008027f9ceb4c81082cb8b234ebd3c1b33c20b1af892L152-R178), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-2e7bd92a74f1fb3fe469008027f9ceb4c81082cb8b234ebd3c1b33c20b1af892L201-R216), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-2e7bd92a74f1fb3fe469008027f9ceb4c81082cb8b234ebd3c1b33c20b1af892L211-R235), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-df2de2d9e49910a87f6432b415e1d2c1b09935c1e21497b304bc9922c20a4b07L36-R51), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eL189-R195), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eL624-R642), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eR649), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eR656), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-adfa337ce44dc2902621da20152a048dac41878cf3716dfc4cc56d03aa212a56R24))
*  Use `tmock` module to mock `Npm` class and its dependencies in `mock-npm` module ([link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-2e7bd92a74f1fb3fe469008027f9ceb4c81082cb8b234ebd3c1b33c20b1af892L59-R75), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-2e7bd92a74f1fb3fe469008027f9ceb4c81082cb8b234ebd3c1b33c20b1af892L97-R109), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-2e7bd92a74f1fb3fe469008027f9ceb4c81082cb8b234ebd3c1b33c20b1af892L226-R247), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-2e7bd92a74f1fb3fe469008027f9ceb4c81082cb8b234ebd3c1b33c20b1af892L254-R267), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-2e7bd92a74f1fb3fe469008027f9ceb4c81082cb8b234ebd3c1b33c20b1af892R278))
*  Use local copy of `nopt` package instead of downloading it from npm registry ([link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519L96-R97))
*  Add `npmcli-docs` package as a dependency of `config` package to generate documentation for config options ([link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-df2de2d9e49910a87f6432b415e1d2c1b09935c1e21497b304bc9922c20a4b07L36-R51), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eR201))
*  Remove unused or deprecated options and files from `config` package ([link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-df2de2d9e49910a87f6432b415e1d2c1b09935c1e21497b304bc9922c20a4b07L27-L28), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-aa3c817f9e6074ccccde6952f62798a646a51b2b3bf76fcfcf2e04abb07afddb), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-74ffdefc33570814788137686f649be840d415feb7f39efa5edf0e4b5abdee5c), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-c8d677d26991f63bf252cb37502de7c1e09f393182294123c6e79358ffd7642f))
*  Update dependency list and hierarchy of `npm` package ([link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-73c2e4c2e886ff07d524b16ddbb70357b31daa755a7fa800fa8dc41140f7544eL780-R809), [link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-7ae45ad102eab3b6d7e7896acd08c427a9b25b346470d7bc6507b6481575d519R198))
*  Update snapshot for `docs` package to reflect changes in `config` package ([link](https://github.com/npm/cli/pull/6020/files?diff=unified&w=0#diff-6f21b9547f08121788002c1b5494afcd450c3d6a9a6f4e1f452d71d2764515b4L689-R689))

### CODE REVIEW

File: lib/utils/config/definitions.js [hunk 1]
```diff
@@ -278,7 +279,7 @@ define('browser', {
   defaultDescription: `
     OS X: \`"open"\`, Windows: \`"start"\`, Others: \`"xdg-open"\`
   `,
-  type: [null, Boolean, String],
+  type: BooleanOrString,
```
Comment (wraithgar) [inline lib/utils/config/definitions.js]: It can still be null right?  That's how the default works?

Comment (lukekarrys) [reply lib/utils/config/definitions.js]: meant to mark this as draft as i was still experimenting with some things, but what I've found so far:

- `default: null` doesn't require that `type` contains `null`
- if `type` contains `null` and not `Boolean` then any `false` value is coerced back to null. so `--no-browser` would end up `null`, which wouldn't break anything currently but i think is a major footgun

Comment (ljharb) [reply lib/utils/config/definitions.js]: what if a user wants to provide a value in a project npmrc to override a user npmrc, with the default? iow, an explicit null

**Comment by ianlewis** [comment]
@wraithgar @lukekarrys Would it be reasonable to break out the `BooleanOrString` validation into another PR in order to fix that cli flags can't be a boolean or string? (though I'm guessing that more may be needed to actually solve the problem).

This PR overall seems to be attempting to be attempting something much broader to improve testing and linting.

**Comment by ianlewis** [comment]
Related #6313

**Comment by wraithgar** [comment]
https://github.com/npm/cli/pull/6314#issuecomment-1553548286

**Pull request closed without merging**